File: Src/Icm.Core/Math/NumberFiles.vb
Imports System.Globalization

Namespace Icm.MathTools
    Public Class NumberFiles

        ''' <summary>
        ''' Reads a simple text file of real numbers with the usual
        ''' (invariant culture) admitted formats. Ignores line comments starting with "#".
        ''' </summary>
        ''' <param name="fn"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Shared Function ReadArrayFile(ByVal fn As String) As Double()
            Dim sr As New System.IO.StreamReader(fn)
            Dim line As String
            Dim l As New Generic.List(Of Double)
            Dim d As Double
            line = sr.ReadLine()

            Do Until line Is Nothing
                If Not line.StartsWith("#", StringComparison.Ordinal) Then
                    If Double.TryParse(line, _
                            NumberStyles.Float, _
                            CultureInfo.InvariantCulture, _
                            d) Then
                        l.Add(d)
                    End If
                End If
                line = sr.ReadLine()
            Loop
            sr.Close()
            Return l.ToArray
        End Function

        ''' <summary>
        '''  Reads a simple file with a multidimensional array of numbers.
        ''' </summary>
        ''' <param name="fn">File name</param>
        ''' <param name="numbersep">Separator for numbers</param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Shared Function ReadMatrixFile(ByVal fn As String, ByVal numbersep As Char) As Double(,)
            Dim sr As New System.IO.StreamReader(fn)
            Dim line As String
            Dim l As New Generic.List(Of Double())
            Dim lineList As New Generic.List(Of Double)
            Dim splitted() As String
            Dim length1 As Integer = 0
            Dim length2 As Integer = 0
            Dim d As Double
            line = sr.ReadLine()

            Do Until line Is Nothing
                If Not line.StartsWith("#", StringComparison.Ordinal) Then
                    lineList.Clear()
                    splitted = line.Split(numbersep)
                    For Each ds As String In splitted
                        If Double.TryParse(ds, _
                            NumberStyles.Float, _
                            CultureInfo.InvariantCulture, _
                            d) Then
                            lineList.Add(d)
                        End If
                    Next
                    If lineList.Count > length2 Then
                        length2 = lineList.Count
                    End If
                    l.Add(lineList.ToArray)
                    length1 += 1
                End If
                line = sr.ReadLine()
            Loop
            sr.Close()

            Dim result = DirectCast(Array.CreateInstance(GetType(Double), length1, length2), Double(,))

            Dim i, j As Integer
            i = 0
            For Each da As Double() In l
                j = 0
                For Each d In da
                    result(i, j) = d
                    j += 1
                Next
                i += 1
            Next
            Return result
        End Function
    End Class
End Namespace

File: Src/Icm.Core/Math/NumberFiles.cs

using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Globalization;

namespace Icm.MathTools
{
	public class NumberFiles
	{

		/// <summary>
		/// Reads a simple text file of real numbers with the usual
		/// (invariant culture) admitted formats. Ignores line comments starting with "#".
		/// </summary>
		/// <param name="fn"></param>
		/// <returns></returns>
		/// <remarks></remarks>
		public static double[] ReadArrayFile(string fn)
		{
			System.IO.StreamReader sr = new System.IO.StreamReader(fn);
			string line = null;
			Generic.List<double> l = new Generic.List<double>();
			double d = 0;
			line = sr.ReadLine();

			while (!(line == null)) {
				if (!line.StartsWith("#", StringComparison.Ordinal)) {
					if (double.TryParse(line, NumberStyles.Float, CultureInfo.InvariantCulture, d)) {
						l.Add(d);
					}
				}
				line = sr.ReadLine();
			}
			sr.Close();
			return l.ToArray;
		}

		/// <summary>
		///  Reads a simple file with a multidimensional array of numbers.
		/// </summary>
		/// <param name="fn">File name</param>
		/// <param name="numbersep">Separator for numbers</param>
		/// <returns></returns>
		/// <remarks></remarks>
		public static double[,] ReadMatrixFile(string fn, char numbersep)
		{
			System.IO.StreamReader sr = new System.IO.StreamReader(fn);
			string line = null;
			Generic.List<double[]> l = new Generic.List<double[]>();
			Generic.List<double> lineList = new Generic.List<double>();
			string[] splitted = null;
			int length1 = 0;
			int length2 = 0;
			double d = 0;
			line = sr.ReadLine();

			while (!(line == null)) {
				if (!line.StartsWith("#", StringComparison.Ordinal)) {
					lineList.Clear();
					splitted = line.Split(numbersep);
					foreach (string ds in splitted) {
						if (double.TryParse(ds, NumberStyles.Float, CultureInfo.InvariantCulture, d)) {
							lineList.Add(d);
						}
					}
					if (lineList.Count > length2) {
						length2 = lineList.Count;
					}
					l.Add(lineList.ToArray);
					length1 += 1;
				}
				line = sr.ReadLine();
			}
			sr.Close();

			dynamic result = (double[,])Array.CreateInstance(typeof(double), length1, length2);

			int i = 0;
			int j = 0;
			i = 0;
			foreach (double[] da in l) {
				j = 0;
				foreach (double d_loopVariable in da) {
					d = d_loopVariable;
					result(i, j) = d;
					j += 1;
				}
				i += 1;
			}
			return result;
		}
	}
}

//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================
